techlibs/microchip/synth_microchip.cc

```diff
@@ -0,0 +1,541 @@
+WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+
+#include "kernel/celltypes.h"
+#include "kernel/log.h"
+#include "kernel/register.h"
+#include "kernel/rtlil.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct SynthMicrochipPass : public ScriptPass {
+	SynthMicrochipPass() : ScriptPass("synth_microchip", "synthesis for Microchip FPGAs") {}
+
+	void help() override
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("    synth_microchip [options]\n");
+		log("\n");
+		log("This command runs synthesis for Microchip FPGAs. Operating on\n");
+		log("partly selected designs is not supported (you must submit a fully-selected \n");
+		log("design). This command creates netlists that are compatible with Microchip \n");
+		log("PolarFire devices.\n");
+		log("\n");
+		log("    -top <module>\n");
+		log("        use the specified module as the top module\n");
+		log("\n");
+		log("    -family <family>\n");
+		log("        Run synthesis for the specified Microchip architecture. \n");
+		log("        Generate the synthesis netlist for the specified family.\n");
+		log("        supported values:\n");
+		log("        - polarfire: PolarFire\n");
+		log("\n");
+		log("    -edif <file>\n");
+		log("        Write the design to the specified edif file. Writing of an output file\n");
+		log("        is omitted if this parameter is not specified.\n");
+		log("\n");
+		log("    -blif <file>\n");
+		log("        Write the design to the specified BLIF file. Writing of an output file\n");
+		log("        is omitted if this parameter is not specified.\n");
+		log("\n");
+		log("    -nobram\n");
+		log("        Do not use block RAM cells in output netlist\n");
+		log("\n");
+		log("    -nocarry\n");
+		log("        Do not use ARI1 cells in output netlist\n");
+		log("\n");
+		log("    -nodsp\n");
+		log("        Do not use MATH blocks to implement multipliers and associated logic\n");
+		log("\n");
+		log("    -noiopad\n");
+		log("        Disable I/O buffer insertion (useful for hierarchical or \n");
+		log("        out-of-context flows)\n");
+		log("\n");
+		log("    -noclkbuf\n");
+		log("        Disable automatic clock buffer insertion\n");
+		log("\n");
+		log("    -run <from_label>:<to_label>\n");
+		log("        Only run the commands between the labels (see below). an empty\n");
+		log("        'from_label' is synonymous to 'begin', and empty 'to_label' is\n");
+		log("        synonymous to the end of the command list.\n");
+		log("\n");
+		log("    -flatten\n");
```

Comment:
This is generally the default, and flows have a `-noflatten` option instead. Xilinx is the odd one out for not flattening by default.